File: BankingSystem.java
import java.util.*;
import java.io.*;

public class BankingSystem {
    private Map<String, User> users;
    private User currentUser;
    private Scanner scanner;
    private static final String DATA_FILE = "bank_data.txt";

    public BankingSystem() {
        users = new HashMap<>();
        scanner = new Scanner(System.in);
        loadData();
    }

    public static void main(String[] args) {
        BankingSystem bank = new BankingSystem();
        bank.run();
    }

    public void run() {
        System.out.println("=== Welcome to Banking System ===");
        
        while (true) {
            if (currentUser == null) {
                showAuthMenu();
            } else {
                showBankingMenu();
            }
        }
    }

    private void showAuthMenu() {
        System.out.println("\n1. Login");
        System.out.println("2. Register");
        System.out.println("3. Exit");
        System.out.print("Choose option: ");
        
        int choice = scanner.nextInt();
        scanner.nextLine(); // consume newline
        
        switch (choice) {
            case 1:
                login();
                break;
            case 2:
                register();
                break;
            case 3:
                saveData();
                System.exit(0);
                break;
            default:
                System.out.println("Invalid option!");
        }
    }

    private void login() {
        System.out.print("Username: ");
        String username = scanner.nextLine();
        System.out.print("Password: ");
        String password = scanner.nextLine();
        
        User user = users.get(username);
        if (user != null && user.getPassword().equals(password)) {
            currentUser = user;
            System.out.println("Login successful!");
        } else {
            System.out.println("Invalid credentials!");
        }
    }

    private void register() {
        System.out.print("Username: ");
        String username = scanner.nextLine();
        
        if (users.containsKey(username)) {
            System.out.println("Username already exists!");
            return;
        }
        
        System.out.print("Password: ");
        String password = scanner.nextLine();
        System.out.print("Email: ");
        String email = scanner.nextLine();
        
        User newUser = new User(username, password, email);
        users.put(username, newUser);
        System.out.println("Registration successful!");
    }

    private void showBankingMenu() {
        System.out.println("\n=== Banking Dashboard ===");
        System.out.println("Current Balance: $" + String.format("%.2f", currentUser.getBalance()));
        System.out.println("\nPress a number key (1-5):");
        System.out.println("1 - Withdraw Amount");
        System.out.println("2 - Deposit Amount");
        System.out.println("3 - Check Balance");
        System.out.println("4 - Total Spent");
        System.out.println("5 - Quit");
        System.out.print("Choose option: ");
        
        int choice = scanner.nextInt();
        
        switch (choice) {
            case 1:
                withdrawAmount();
                break;
            case 2:
                depositAmount();
                break;
            case 3:
                checkBalance();
                break;
            case 4:
                showTotalSpent();
                break;
            case 5:
                quit();
                break;
            default:
                System.out.println("Invalid option! Please press 1-5.");
        }
    }

    private void withdrawAmount() {
        System.out.print("Enter amount to withdraw: $");
        double amount = scanner.nextDouble();
        
        if (amount <= 0) {
            System.out.println("Invalid amount!");
            return;
        }
        
        if (amount > currentUser.getBalance()) {
            System.out.println("Insufficient balance!");
            return;
        }
        
        currentUser.withdraw(amount);
        System.out.println("Successfully withdrew $" + String.format("%.2f", amount));
        System.out.println("New balance: $" + String.format("%.2f", currentUser.getBalance()));
    }

    private void depositAmount() {
        System.out.print("Enter amount to deposit: $");
        double amount = scanner.nextDouble();
        
        if (amount <= 0) {
            System.out.println("Invalid amount!");
            return;
        }
        
        currentUser.deposit(amount);
        System.out.println("Successfully deposited $" + String.format("%.2f", amount));
        System.out.println("New balance: $" + String.format("%.2f", currentUser.getBalance()));
    }

    private void checkBalance() {
        System.out.println("Current Balance: $" + String.format("%.2f", currentUser.getBalance()));
    }

    private void showTotalSpent() {
        System.out.println("Total Spent: $" + String.format("%.2f", currentUser.getTotalSpent()));
    }

    private void quit() {
        System.out.println("Thank you for using Banking System!");
        saveData();
        currentUser = null;
    }

    private void saveData() {
        try (PrintWriter writer = new PrintWriter(new FileWriter(DATA_FILE))) {
            for (User user : users.values()) {
                writer.println(user.getUsername() + "," + user.getPassword() + "," + 
                             user.getEmail() + "," + user.getBalance() + "," + user.getTotalSpent());
            }
        } catch (IOException e) {
            System.out.println("Error saving data: " + e.getMessage());
        }
    }

    private void loadData() {
        try (BufferedReader reader = new BufferedReader(new FileReader(DATA_FILE))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length == 5) {
                    User user = new User(parts[0], parts[1], parts[2]);
                    user.setBalance(Double.parseDouble(parts[3]));
                    user.setTotalSpent(Double.parseDouble(parts[4]));
                    users.put(parts[0], user);
                }
            }
        } catch (IOException e) {
            // File doesn't exist yet, that's okay
        }
    }
}

class User {
    private String username;
    private String password;
    private String email;
    private double balance;
    private double totalSpent;

    public User(String username, String password, String email) {
        this.username = username;
        this.password = password;
        this.email = email;
        this.balance = 1000.0; // Starting balance
        this.totalSpent = 0.0;
    }

    public void withdraw(double amount) {
        balance -= amount;
        totalSpent += amount;
    }

    public void deposit(double amount) {
        balance += amount;
    }

    // Getters and Setters
    public String getUsername() { return username; }
    public String getPassword() { return password; }
    public String getEmail() { return email; }
    public double getBalance() { return balance; }
    public double getTotalSpent() { return totalSpent; }
    
    public void setBalance(double balance) { this.balance = balance; }
    public void setTotalSpent(double totalSpent) { this.totalSpent = totalSpent; }
}

File: banking_system.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_USERS 100
#define MAX_USERNAME 50
#define MAX_PASSWORD 50
#define MAX_EMAIL 100
#define DATA_FILE "bank_data_c.txt"

typedef struct {
    char username[MAX_USERNAME];
    char password[MAX_PASSWORD];
    char email[MAX_EMAIL];
    double balance;
    double totalSpent;
} User;

typedef struct {
    User users[MAX_USERS];
    int userCount;
    int currentUserIndex;
} BankingSystem;

BankingSystem bank;

void initializeSystem();
void saveData();
void loadData();
void showAuthMenu();
void showBankingMenu();
void login();
void registerUser();
void withdrawAmount();
void depositAmount();
void checkBalance();
void showTotalSpent();
void quit();
int findUser(const char* username);
void clearInputBuffer();

int main() {
    initializeSystem();
    loadData();
    
    printf("=== Welcome to Banking System ===\n");
    
    while (1) {
        if (bank.currentUserIndex == -1) {
            showAuthMenu();
        } else {
            showBankingMenu();
        }
    }
    
    return 0;
}

void initializeSystem() {
    bank.userCount = 0;
    bank.currentUserIndex = -1;
}

void showAuthMenu() {
    int choice;
    
    printf("\n1. Login\n");
    printf("2. Register\n");
    printf("3. Exit\n");
    printf("Choose option: ");
    
    scanf("%d", &choice);
    clearInputBuffer();
    
    switch (choice) {
        case 1:
            login();
            break;
        case 2:
            registerUser();
            break;
        case 3:
            saveData();
            exit(0);
            break;
        default:
            printf("Invalid option!\n");
    }
}

void login() {
    char username[MAX_USERNAME];
    char password[MAX_PASSWORD];
    
    printf("Username: ");
    fgets(username, sizeof(username), stdin);
    username[strcspn(username, "\n")] = 0; // Remove newline
    
    printf("Password: ");
    fgets(password, sizeof(password), stdin);
    password[strcspn(password, "\n")] = 0; // Remove newline
    
    int userIndex = findUser(username);
    if (userIndex != -1 && strcmp(bank.users[userIndex].password, password) == 0) {
        bank.currentUserIndex = userIndex;
        printf("Login successful!\n");
    } else {
        printf("Invalid credentials!\n");
    }
}

void registerUser() {
    if (bank.userCount >= MAX_USERS) {
        printf("Maximum users reached!\n");
        return;
    }
    
    char username[MAX_USERNAME];
    char password[MAX_PASSWORD];
    char email[MAX_EMAIL];
    
    printf("Username: ");
    fgets(username, sizeof(username), stdin);
    username[strcspn(username, "\n")] = 0;
    
    if (findUser(username) != -1) {
        printf("Username already exists!\n");
        return;
    }
    
    printf("Password: ");
    fgets(password, sizeof(password), stdin);
    password[strcspn(password, "\n")] = 0;
    
    printf("Email: ");
    fgets(email, sizeof(email), stdin);
    email[strcspn(email, "\n")] = 0;
    
    // Create new user
    strcpy(bank.users[bank.userCount].username, username);
    strcpy(bank.users[bank.userCount].password, password);
    strcpy(bank.users[bank.userCount].email, email);
    bank.users[bank.userCount].balance = 1000.0;
    bank.users[bank.userCount].totalSpent = 0.0;
    
    bank.userCount++;
    printf("Registration successful!\n");
}

void showBankingMenu() {
    int choice;
    User* currentUser = &bank.users[bank.currentUserIndex];
    
    printf("\n=== Banking Dashboard ===\n");
    printf("Current Balance: $%.2f\n", currentUser->balance);
    printf("\nPress a number key (1-5):\n");
    printf("1 - Withdraw Amount\n");
    printf("2 - Deposit Amount\n");
    printf("3 - Check Balance\n");
    printf("4 - Total Spent\n");
    printf("5 - Quit\n");
    printf("Choose option: ");
    
    scanf("%d", &choice);
    
    switch (choice) {
        case 1:
            withdrawAmount();
            break;
        case 2:
            depositAmount();
            break;
        case 3:
            checkBalance();
            break;
        case 4:
            showTotalSpent();
            break;
        case 5:
            quit();
            break;
        default:
            printf("Invalid option! Please press 1-5.\n");
    }
}

void withdrawAmount() {
    double amount;
    User* currentUser = &bank.users[bank.currentUserIndex];
    
    printf("Enter amount to withdraw: $");
    scanf("%lf", &amount);
    
    if (amount <= 0) {
        printf("Invalid amount!\n");
        return;
    }
    
    if (amount > currentUser->balance) {
        printf("Insufficient balance!\n");
        return;
    }
    
    currentUser->balance -= amount;
    currentUser->totalSpent += amount;
    
    printf("Successfully withdrew $%.2f\n", amount);
    printf("New balance: $%.2f\n", currentUser->balance);
}

void depositAmount() {
    double amount;
    User* currentUser = &bank.users[bank.currentUserIndex];
    
    printf("Enter amount to deposit: $");
    scanf("%lf", &amount);
    
    if (amount <= 0) {
        printf("Invalid amount!\n");
        return;
    }
    
    currentUser->balance += amount;
    
    printf("Successfully deposited $%.2f\n", amount);
    printf("New balance: $%.2f\n", currentUser->balance);
}

void checkBalance() {
    User* currentUser = &bank.users[bank.currentUserIndex];
    printf("Current Balance: $%.2f\n", currentUser->balance);
}

void showTotalSpent() {
    User* currentUser = &bank.users[bank.currentUserIndex];
    printf("Total Spent: $%.2f\n", currentUser->totalSpent);
}

void quit() {
    printf("Thank you for using Banking System!\n");
    saveData();
    bank.currentUserIndex = -1;
}

int findUser(const char* username) {
    for (int i = 0; i < bank.userCount; i++) {
        if (strcmp(bank.users[i].username, username) == 0) {
            return i;
        }
    }
    return -1;
}

void saveData() {
    FILE* file = fopen(DATA_FILE, "w");
    if (file == NULL) {
        printf("Error saving data!\n");
        return;
    }
    
    for (int i = 0; i < bank.userCount; i++) {
        fprintf(file, "%s,%s,%s,%.2f,%.2f\n",
                bank.users[i].username,
                bank.users[i].password,
                bank.users[i].email,
                bank.users[i].balance,
                bank.users[i].totalSpent);
    }
    
    fclose(file);
}

void loadData() {
    FILE* file = fopen(DATA_FILE, "r");
    if (file == NULL) {
        return; // File doesn't exist yet
    }
    
    char line[256];
    while (fgets(line, sizeof(line), file) && bank.userCount < MAX_USERS) {
        char* token = strtok(line, ",");
        if (token) strcpy(bank.users[bank.userCount].username, token);
        
        token = strtok(NULL, ",");
        if (token) strcpy(bank.users[bank.userCount].password, token);
        
        token = strtok(NULL, ",");
        if (token) strcpy(bank.users[bank.userCount].email, token);
        
        token = strtok(NULL, ",");
        if (token) bank.users[bank.userCount].balance = atof(token);
        
        token = strtok(NULL, ",");
        if (token) bank.users[bank.userCount].totalSpent = atof(token);
        
        bank.userCount++;
    }
    
    fclose(file);
}

void clearInputBuffer() {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}